File: Tallest_Billboard.java
class Solution {
    public int tallestBillboard(int[] rods) {
        Map<Integer, Integer> dp = new HashMap<>();
        dp.put(0, 0);

        for (int rod : rods) {
            Map<Integer, Integer> currDP = new HashMap<>(dp);

            for (Map.Entry<Integer, Integer> entry : dp.entrySet()) {
                int diff = entry.getKey();
                int highest = entry.getValue();

                int leftDiff = diff + rod;
                currDP.put(leftDiff, Math.max(currDP.getOrDefault(leftDiff, 0), highest + rod));

                int rightRod = highest - diff + rod;
                int rightDiff = Math.abs(highest - rightRod);
                currDP.put(rightDiff, Math.max(currDP.getOrDefault(rightDiff, 0), Math.max(highest, rightRod)));
            }

            dp = currDP;
        }

        return dp.get(0);        
    }
}

File: Tallest_Billboard.js
/**
 * @param {number[]} rods
 * @return {number}
 */
var tallestBillboard = function(rods) {
    const dp = {0:0};

    for (const rod of rods) {
        const dpEntries = Object.entries(dp);

        for (const [diff, highest] of dpEntries) {
            const leftDiff = parseInt(diff) + rod;
            dp[leftDiff] = Math.max(dp[leftDiff] || 0, highest + rod);

            const rightRod = highest - parseInt(diff) + rod;
            const rightDiff = Math.abs(highest - rightRod);
            dp[rightDiff] = Math.max(dp[rightDiff] || 0, Math.max(highest, rightRod));
        }
    }

    return dp[0];    
};
